powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/SignatureUtils.java

```diff
@@ -83,46 +85,118 @@ public boolean validateECDSASignature(byte[] signedBytes, byte[] signature, Publ
+        final HMACHashUtilities hmac = new HMACHashUtilities();
 
-        // Prepare holder for signature components
-        String[] signatureComponents = new String[signatureKeys.size()];
+        // Prepare array for signature binary components.
+        final byte[][] signatureComponents = new byte[signatureKeys.size()][];
 
-        CryptoProviderUtil keyConvertor = PowerAuthConfiguration.INSTANCE.getKeyConvertor();
+        final CryptoProviderUtil keyConverter = PowerAuthConfiguration.INSTANCE.getKeyConvertor();
 
         for (int i = 0; i < signatureKeys.size(); i++) {
-            byte[] signatureKey = keyConvertor.convertSharedSecretKeyToBytes(signatureKeys.get(i));
+            final byte[] signatureKey = keyConverter.convertSharedSecretKeyToBytes(signatureKeys.get(i));
             byte[] derivedKey = hmac.hash(signatureKey, ctrData);
 
             for (int j = 0; j < i; j++) {
-                byte[] signatureKeyInner = keyConvertor.convertSharedSecretKeyToBytes(signatureKeys.get(j + 1));
+                byte[] signatureKeyInner = keyConverter.convertSharedSecretKeyToBytes(signatureKeys.get(j + 1));
                 byte[] derivedKeyInner = hmac.hash(signatureKeyInner, ctrData);
                 derivedKey = hmac.hash(derivedKeyInner, derivedKey);
             }
 
-            byte[] signatureLong = hmac.hash(derivedKey, data);
-
-            if (signatureLong.length < 4) { // assert
+            final byte[] signatureBytes = hmac.hash(derivedKey, data);
+            if (signatureBytes.length < PowerAuthConfiguration.SIGNATURE_BINARY_LENGTH) { // assert
                 throw new IndexOutOfBoundsException();
             }
-            int index = signatureLong.length - 4;
-            int number = (ByteBuffer.wrap(signatureLong).getInt(index) & 0x7FFFFFFF) % (int) (Math.pow(10, PowerAuthConfiguration.SIGNATURE_LENGTH));
-            String signature = String.format("%0" + PowerAuthConfiguration.SIGNATURE_LENGTH + "d", number);
-            signatureComponents[i] = signature;
+            signatureComponents[i] = signatureBytes;
         }
 
-        return Joiner.on("-").join(signatureComponents);
+        return signatureComponents;
+    }
+
+    /**
+     * Compute PowerAuth signature for given data using a secret signature keys and counter byte array.
+     *
+     * @param data Data to be signed.
+     * @param signatureKeys Keys for computing the signature.
+     * @param ctrData Counter byte array / derived key index.
+     * @param format Format of signature to produce.
+     * @return PowerAuth signature for given data.
+     * @throws GenericCryptoException In case signature computation fails.
+     * @throws CryptoProviderException In case cryptography provider is incorrectly initialized.
+     */
+    public String computePowerAuthSignature(byte[] data, List<SecretKey> signatureKeys, byte[] ctrData, PowerAuthSignatureFormat format) throws GenericCryptoException, CryptoProviderException {
```

Comment:
Shouldn't we check input parameters in this method? I know this was not done before, but it is a good practice. We can check for nullability of all parameters and make sure `signatureKeys` list is not empty.

Reply:
Right, but I think that it's OK to have `null` data. That's acceptable in HMAC calculation and means an empty data.

Reply:
Yes, but `null` `ctrData`, `format` or `signatureKeys` is not so cool.